core/toolbox/category.js

```diff
@@ -34,612 +34,628 @@ const {ToolboxItem} = goog.require('Blockly.ToolboxItem');
+     * @type {?Element}
+     * @protected
+     */
+    this.labelDom_ = null;
+
+    /**
+     * All the css class names that are used to create a category.
+     * @type {!ToolboxCategory.CssConfig}
+     * @protected
+     */
+    this.cssConfig_ = this.makeDefaultCssConfig_();
+
+    /**
+     * True if the category is meant to be hidden, false otherwise.
+     * @type {boolean}
+     * @protected
+     */
+    this.isHidden_ = false;
+
+    /**
+     * True if this category is disabled, false otherwise.
+     * @type {boolean}
+     * @protected
+     */
+    this.isDisabled_ = false;
+
+    /**
+     * The flyout items for this category.
+     * @type {string|!toolbox.FlyoutItemInfoArray}
+     * @protected
+     */
+    this.flyoutItems_ = [];
+  }
 
   /**
-   * The name that will be displayed on the category.
-   * @type {string}
+   * Creates an object holding the default classes for a category.
+   * @return {!ToolboxCategory.CssConfig} The configuration object holding
+   *    all the CSS classes for a category.
    * @protected
    */
-  this.name_ = parsing.replaceMessageReferences(categoryDef['name']);
+  makeDefaultCssConfig_() {
+    return {
+      'container': 'blocklyToolboxCategory',
+      'row': 'blocklyTreeRow',
+      'rowcontentcontainer': 'blocklyTreeRowContentContainer',
+      'icon': 'blocklyTreeIcon',
+      'label': 'blocklyTreeLabel',
+      'contents': 'blocklyToolboxContents',
+      'selected': 'blocklyTreeSelected',
+      'openicon': 'blocklyTreeIconOpen',
+      'closedicon': 'blocklyTreeIconClosed',
+    };
+  }
 
   /**
-   * The colour of the category.
-   * @type {string}
+   * Parses the contents array depending on if the category is a dynamic
+   * category, or if its contents are meant to be shown in the flyout.
+   * @param {!toolbox.CategoryInfo} categoryDef The information needed
+   *     to create a category.
    * @protected
    */
-  this.colour_ = this.getColour_(categoryDef);
+  parseContents_(categoryDef) {
+    const contents = categoryDef['contents'];
+
+    if (categoryDef['custom']) {
+      this.flyoutItems_ = categoryDef['custom'];
+    } else if (contents) {
+      for (let i = 0; i < contents.length; i++) {
+        const itemDef = contents[i];
+        const flyoutItem =
+            /** @type {toolbox.FlyoutItemInfo} */ (itemDef);
+        this.flyoutItems_.push(flyoutItem);
+      }
+    }
+  }
 
   /**
-   * The html container for the category.
-   * @type {?Element}
+   * Parses the non-contents parts of the category def.
+   * @param {!toolbox.CategoryInfo} categoryDef The information needed to create
+   *     a category.
    * @protected
    */
-  this.htmlDiv_ = null;
+  parseCategoryDef_(categoryDef) {
+    this.name_ = parsing.replaceMessageReferences(categoryDef['name']);
+    this.colour_ = this.getColour_(categoryDef);
+    object.mixin(
+        this.cssConfig_, categoryDef['cssconfig'] || categoryDef['cssConfig']);
+  }
+
+  /**
+   * @override
```

Comment:
Could you add a comment to this method saying it must be called immediately after creation? The super class's `init` is a no-op so it's good to be explicit here. Also I would suggest moving this up to be the first method after the constructor so it's easier to notice.